### PR TITLE
feat(SideNavMenu): collapse submenu on esc keydown

### LINKED ISSUE #3590: AVT 2 & 3 - React Header Base w/SideNav, Fixed SideNav, & SideNav w/Icon UI Shell has keyboard navigation issues
## Environment
macOS Mojave version 10.14.5
Chrome Version 75.0.3770.100 (Official Build) (64-bit)
Carbon v10 - React
DAP - July 2019 Ruleset

## Detailed Description
Tab to the side menu and use up and down arrows to navigate the menu and nothing happens.
Esc does not close the menu.
**Expected result:** Up and down arrow keys should navigate up and down the menu. Esc key should close the menu. 

See [ARIA Authoring Practices for menu and menubar](https://www.w3.org/TR/wai-aria-practices-1.1/#menu) for keyboard navigation.

### PR DESCRIPTION
**Pull request by emyarod**
Closes #3590

This PR adds <kbd>Esc</kbd> key support to side nav menus so that they can be collapsed via keyboard only

#### Changelog

**New**

- `handleKeydown` method

#### Testing / Reviewing

Ensure that the <kbd>Esc</kbd> key functions as expected while tabbing through UI shell side nav menus

### CODE REVIEW

File: packages/react/src/components/UIShell/__tests__/SideNavMenu-test.js [hunk 1]
```diff
@@ -43,6 +43,17 @@ describe('SideNavMenu', () => {
     expect(wrapper.state('isExpanded')).toBe(true);
   });
 
+  fit('should collapse the menu when the Esc key is pressed', () => {
```
Comment (abbeyhrt) [inline packages/react/src/components/UIShell/__tests__/SideNavMenu-test.js]:
```suggestion
  it('should collapse the menu when the Esc key is pressed', () => {
```

**Review by tw15egan — APPROVED** [review]
Working as expected 👍 ✅

**Comment by joshblack** [comment]
Is the up/down arrow key support already in the component btw? Just was looking at the linked issue 👀

**Review by asudoh — APPROVED** [review]
LGTM 👍 - Thanks @emyarod!

Update: As #3590 states, I think this PR focuses on <kbd>Esc</kbd> key support.

**Comment by emyarod** [comment]
unless arrow key nav is a strict requirement, <kbd>Tab</kbd>, <kbd>Shift + Tab</kbd>, <kbd>Space</kbd>, and <kbd>Enter</kbd> are functional for navigation and toggling menu state. this PR will add <kbd>Esc</kbd> key support

**Comment by joshblack** [comment]
I'm fine either way, just was asking since the related issue listed it as a bug and wanted to see if it was being addressed in the PR. Do we know if it's a requirement? If so, we might not want the PR to close that issue. If not, then definitely makes sense to close it 👍

**Comment by emyarod** [comment]
there were a few similar old issues that referenced lack of keyboard navigation which I closed earlier this week since it looks like we now support keyboard navigation, just not the arrow key navigation specifically. but if arrow key nav is a strict requirement we can also reopen the issue

**Comment by joshblack** [comment]
@emyarod not sure which issue you're referring too, I'm just looking at this line from #3590:

> Expected result: Up and down arrow keys should navigate up and down the menu. Esc key should close the menu.

It seems like the expectation is that this follows the menu and menubar pattern, to your point I don't think we're doing that, right? cc @dakahn do you know what the true requirements are here?

**Comment by emyarod** [comment]
@joshblack I was referring to previous tickets related to the topic including the old epic created by @dakahn which mentions not strictly adhering to the menu keyboard nav guidelines

https://github.com/carbon-design-system/carbon/issues/4766 https://github.com/carbon-design-system/carbon/issues/3763

**Comment by joshblack** [comment]
Cool! Sounds good to me then 👍 Thanks!

**Comment by dakahn** [comment]
Yeah that menu pattern is controversial and introduces complications (menu being a system level designation vs the list of links or dropdowns found on the web).

Here's a cool primer
https://adrianroselli.com/2017/10/dont-use-aria-menu-roles-for-site-nav.html

**Pull request merged**